sstables/sstables.cc

```diff
@@ -1410,6 +1410,66 @@ void sstable::write_filter() {
+    // false positive rate.
+    auto curr_bitset_size = static_cast<utils::filter::bloom_filter*>(_components->filter.get())->bits().memory_size();
+    auto bitset_size_lower_bound = utils::i_filter::get_filter_size(num_partitions,
+                                                                    _schema->bloom_filter_fp_chance() * 1.25);
+    auto bitset_size_upper_bound = utils::i_filter::get_filter_size(num_partitions,
+                                                                    _schema->bloom_filter_fp_chance() * 0.75);
+    if (bitset_size_lower_bound <= curr_bitset_size && curr_bitset_size <= bitset_size_upper_bound) {
+        return;
+    }
+
+    // Consumer that adds the keys from index entries to the given bloom filter
+    class bloom_filter_builder {
+        utils::filter_ptr& _filter;
+    public:
+        bloom_filter_builder(utils::filter_ptr &filter) : _filter(filter) {}
+        void consume_entry(parsed_partition_index_entry&& e) {
+            _filter->add(to_bytes_view(e.key));
+        }
+    };
+
+    sstlog.info("Rebuilding bloom filter {}", filename(component_type::Filter));
+
+    // Replace the existing filter with a new one that can optimally represent the given num_partitions.
+    // The rebuilding is done in-place as this method is only called before the sstable is sealed.
+    _components->filter = utils::i_filter::get_filter(num_partitions, _schema->bloom_filter_fp_chance(), get_filter_format(_version));
+
+    auto index_file = open_file(component_type::Index, open_flags::ro).get();
+    auto index_file_closer = deferred_action([&index_file] {
+        try {
+            index_file.close().get();
+        } catch (...) {
+            sstlog.warn("sstable close index_file failed: {}", std::current_exception());
+            general_disk_error();
+        }
+    });
+    auto index_file_size = index_file.size().get();
+
+    auto sem = reader_concurrency_semaphore(reader_concurrency_semaphore::no_limits{}, "sstable::rebuild_filter_from_index",
+            reader_concurrency_semaphore::register_metrics::no);
+    auto sem_stopper = deferred_stop(sem);
+
+    // rebuild the filter using index_consume_entry_context
+    auto bfb_consumer = bloom_filter_builder(_components->filter);
+    auto consumer_ctx = make_lw_shared<index_consume_entry_context<bloom_filter_builder>>(
+            *this, sem.make_tracking_only_permit(_schema, "rebuild_filter_from_index", db::no_timeout, {}), bfb_consumer, trust_promoted_index::no,
+            make_file_input_stream(index_file, 0, index_file_size, {.buffer_size = sstable_buffer_size}), 0, index_file_size,
+            (_version >= sstable_version_types::mc
+                ? std::make_optional(get_clustering_values_fixed_lengths(get_serialization_header()))
+                : std::optional<column_values_fixed_lengths>{}));
+    auto consumer_ctx_closer = deferred_close(*consumer_ctx);
+    consumer_ctx->consume_input().get();
```

Comment:
Something that's missing is a way to abort if we rebuild the filter during shutdown.

IIRC all the interrutibility checks are outside the sstable writer. @raphaelsc please suggest something.

Reply:
@lkshminarayanan we achieve it here by holding abort_source (from stop_signal in main.cc) in sstables_manager and forwarding it here. if abort is requested, bloom_filter_builder triggers an exception aborting the sstable write and table is prepared for not retrying it on shutdown (so is compaction manager). similar approach is done for sstable validate that uses read machinery only.

Reply:
I am thinking if sstable readers themselves should become abortable (we have ability to cancel IO intent too), but we don't have to do that now. It's ok to abort on partition boundary here as we skip promoted indexes.

Reply:
@raphaelsc I have created a separate PR to enable aborting from inside the index consumer. https://github.com/scylladb/scylladb/pull/19453

Depending on which gets merged first, I'll rebase the other on top of that.